aten/src/ATen/core/jit_type.h

```diff
@@ -1266,17 +1268,36 @@ CAFFE2_API bool isSubvalueOf(const IValue& input_ivalue, TypePtr type);
 
 using TypeEnv = std::unordered_map<std::string, TypePtr>;
 struct MatchTypeReturn {
-  MatchTypeReturn(TypePtr type) : type(type) {}
-  MatchTypeReturn(std::string errMsg) : errMsg(std::move(errMsg)) {}
+  MatchTypeReturn(std::string reason) : reason_(std::move(reason)) {}
+  static MatchTypeReturn Success() {
+    return MatchTypeReturn();
+  }
+  bool success() const {
+    return reason_ == "";
```

Comment:
Make `reason_` an optional to avoid false success if you report a failure with an empty string for a reason?